Use chromium's side panel inside our side bar container, and insert Reading List

petemill opened the pull request:
Chromium's side panel becomes controlled by Brave's side bar buttons, which uses chromium's side panel coordinator.

The goal of this PR is to get the best of both worlds - use Chromium's SidePanel related views, models, coordinators, WebContents management, and WebUI resources, whilst still using the awesome Brave SideBar from @simonhong.

There are many things we and chromium would like to use side panels for, and this could set us up to scale well for that.

Another change is that using the toolbar button to open side panel will result in the sidebar (and panel) staying open, even if the sidebar prefs are set to open on mouseover or never. Once the panel closes, then the sidebar will to if the prefs have either of these values.

_Note: One consideration is to instead do the inverse of what I have done in this PR: Perhaps we move our SideBar to be a child of `browser_view_->right_aligned_side_panel()` (like the ComboBox is for Chromium) and have the toolbar button turn on SideBar + SidePanel, with no separation between the two. Some complication will need to be catered to where we want the SideBar to show even when there is no active SidePanel._

### TODO
- [x] Use Chromium's SidePanelCoordinator to manage WebContents and panel open / close
- [x] Reading list works
- [x] Bookmarks works, and remove Brave's clone of the panel
- [x] Panel opens and closes when clicking SideBar item
- [x] Design team feedback on padding and border colors
- [ ] Panel gets focus when opening
- [x] Toolbar button controls SideBar, not just panel inside it
- [x] Replace toolbar button icon so that sidebar is on left
- [ ] Use Brave design styles for Reading list panel, not chromium (for follow-up)
- [ ] Have side panel show on right for RTL UI (for follow-up)

<img width="1402" alt="image" src="https://user-images.githubusercontent.com/741836/178616073-7ff28568-6fb6-479d-a4f1-06d73f068c0d.png">


<!-- Add brave-browser issue bellow that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/17959

### Future considerations:
- [ ] Side search
- [ ] Interaction with vertical tabs, and perhaps Side Bar + Side Panel moving to the right

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
On https://github.com/brave/brave-browser/issues/17959

simonhong commented on browser/ui/views/frame/brave_browser_view.cc:
```diff
@@ -167,11 +167,15 @@ BraveBrowserView::BraveBrowserView(std::unique_ptr<Browser> browser)
+      std::make_unique<SidebarContainerView>(GetBraveBrowser(),
+                                             side_panel_coordinator(),
+                                             std::move(original_side_panel)));
+  right_aligned_side_panel_ = sidebar_container_view_->side_panel();
```
This replacement seems great! 👍🏼

simonhong commented on browser/ui/views/sidebar/sidebar_container_view.cc:
```diff
@@ -179,8 +158,8 @@ void SidebarContainerView::UpdateBackgroundAndBorder() {
-  sidebar_panel_webview_ =
-      AddChildView(std::make_unique<SidebarPanelWebView>(browser_->profile()));
+
+  ReorderChildView(side_panel_, 1);
```
If we add `sidebar_control_view_` at index `0`, then maybe we don't need to reorder `side_pane_`?
```
  sidebar_control_view_ =
      AddChildViewAt(std::make_unique<SidebarControlView>(browser_), 0);
```

simonhong reviewed (commented):
Great to reuse upstream's side panel UI 👍🏼

petemill commented on components/sidebar/constants.h:
```diff
@@ -14,12 +14,7 @@ constexpr char kSidebarItemBuiltInItemTypeKey[] = "built_in_item_type";
 constexpr char kSidebarItemOpenInPanelKey[] = "open_in_panel";
 
-// TODO(simonhong): Move this to
-// //brave/components/constants/webui_url_constants.h when default builtin items
```
That constants file is deleted - not sure if this still applies?

simonhong replied:
Safe to delete as this file doesn't have copied url from chrome layer. 👍🏼

petemill commented on browser/ui/views/sidebar/sidebar_container_view.cc:
```diff
@@ -37,22 +38,24 @@
+    default:
+      // We really shouldn't get here, add a new case for any new types which
+      // we want to support.
+      return SidePanelEntry::Id::kReadingList;
```
Yes I think so, especially if it's WebUI

petemill commented:
I have now pushed all functionality working well with the new toolbar icon. The Sidebar correctly shows active states when the panel is made active from external means (i.e. the side panel toolbar icon).

Another change is that using the toolbar button to open side panel will result in the sidebar (and panel) staying open, even if the sidebar prefs are set to open on mouseover or never. Once the panel closes, then the sidebar will to if the prefs have either of these values.

The only item left for me is to update the vector icon from figma.

petemill commented:
Test plans added to https://github.com/brave/brave-browser/issues/17959

petemill commented:
All feedback addressed @simonhong

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/chromium-side-panel/aa7ef882e7a996f7fb477940088c147362f692e6/storybook/index.html)

simonhong commented:
I realized that reading list item is not added to existing uesrs.
`SidebarService` doesn't show newly introduced default item.
It should be handled as a separated issue.

simonhong commented:
When any item is not activated after launching, clicking toolbar button opens always reading list.
When readling list item is removed, I think another (bookmark) item should be activated.
If there are no visible items for panel, what should we do when clicks toolbar button?

also there is no border between web contens and panel when panel is opened via toolbar button after launching.
<img width="387" alt="image" src="https://user-images.githubusercontent.com/6786187/178727606-248c56e5-296e-477d-ad36-9ccc2505c710.png">

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/chromium-side-panel/71b412ce8bc8998a0d97bec5fd611055fb871324/storybook/index.html)

petemill commented:
@simonhong I've addressed feedback again and added / amended some tests as appropriate. Please re-review when you can.

> I realized that SidebarService doesn't show newly introduced default item.
So existing user should add reading list item via add item bubble.
It should be handled as a separated issue.

I've handled that in this PR now. I migrated the prefs such that the existing pref only stores user items (not built-in). And I made a new pref which simply contains a list of IDs of built-in items which the user has hidden.

> When any item is not activated after launching, clicking toolbar button opens always reading list (even it's removed item).
When readling list item is removed, I think another (bookmark) item should be activated.
If there are no visible items for panel, what should we do when clicks toolbar button?

This should now work as you describe. And if the user removes all panel items, then the toolbar icon will hide.

> also there is no border between web contens and panel when panel is opened via toolbar button after launching.

This is fixed. The issue was the delay between sidebar item getting activated and sidepanel view being shown (due to waiting for load event).

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/chromium-side-panel/691bc97ca054daa679b5dfaaefd02b7f00c5ff1d/storybook/index.html)

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/chromium-side-panel/c851e0763a4d4b71ebea2099fec2d52a338af44e/storybook/index.html)

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/chromium-side-panel/918e874923ba7d9a134df6ef41ed9f8bc06897f6/storybook/index.html)

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/chromium-side-panel/785b41f54e97d3ee0875bc4473257000f6b30f10/storybook/index.html)

sangwoo108 commented on components/sidebar/sidebar_service.cc:
```diff
@@ -225,75 +350,133 @@ SidebarService::ShowSidebarOption SidebarService::GetSidebarShowOption() const {
+            continue;
+          }
+          // Valid built-in item, add it
+          items_.emplace_back(*std::make_move_iterator(iter));
```
I think moving this way looks really great but If we apply NoDestructor() and const& to GetDefaultSidebarItems(), this may not work 😥

petemill replied:
I don't see a suggestion to apply const& to `GetDefaultSidebarItems()`, just `GetDefaultBuiltInItemTypes()` ;-) So for now we are ok to keep this style, and we can debate if `GetDefaultSidebarItems()` should be const another time.

sangwoo108 replied:
Oh, yes. I was mistaken. Thanks!

sangwoo108 reviewed (commented):
Nice work!🙌

Sorry for being picky. Please feel free to give me feedbacks for my comments.

mkarolin commented on chromium_src/chrome/browser/ui/views/side_search/side_search_browser_controller.cc:
```diff
@@ -0,0 +1,14 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Need to include this header here otherwise SidePanel macro gets undef'ed
```
Typically we include the corresponding header first to prevent the redefinitions. Should
#include "chrome/browser/ui/views/side_search/side_search_browser_controller.h" here?

petemill replied:
thanks!

petemill replied:
Seems like we also need to keep the browser_view.h include as that also has a macro called SidePanel

mkarolin commented on chromium_src/chrome/browser/ui/views/frame/browser_view.cc:
```diff
@@ -3,16 +3,24 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this file,
  * You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+#include "brave/browser/ui/views/frame/brave_browser_view_layout.h"
```
Should include "chrome/browser/ui/views/frame/browser_view.h" first?

simonhong commented on browser/ui/sidebar/sidebar_controller.cc:
```diff
@@ -134,8 +135,6 @@ void SidebarController::LoadAtTab(const GURL& url) {
 void SidebarController::OnShowSidebarOptionChanged(
     SidebarService::ShowSidebarOption option) {
-  // Clear active state whenever sidebar enabled state is changed.
-  ActivateItemAt(-1);
```
My assumption is `ActivateItemAt(-1);` was deleted to make sidebar panel(+sidebar control view) open till user closes opened panel explicitely even if user changed the show option. @petemill Am I right?

But, sidebar control view doesn't get hidden after `never` option is set via context menu even if panel is not opened. I think this is reggression.

simonhong replied:
I found the cause of this regression and seems fixed with below changes.
`ShouldShowSidebar()` returns true because context menu was not hidden.
```
diff --git a/browser/ui/views/sidebar/sidebar_container_view.cc b/browser/ui/views/sidebar/sidebar_container_view.cc
index 9e62f62627..2a03f8f3b8 100644
--- a/browser/ui/views/sidebar/sidebar_container_view.cc
+++ b/browser/ui/views/sidebar/sidebar_container_view.cc
@@ -125,11 +125,11 @@ void SidebarContainerView::SetSidebarShowOption(
   }
   // We don't set false explicitly just in case a panel view is still open.
   if (show_option == ShowSidebarOption::kShowNever) {
-    ShowSidebar(ShouldShowSidebar(), false);
+    ShowSidebar(IsPanelActive(), false);
     return;
   }
 
-  ShowSidebar(ShouldShowSidebar(), true);
+  ShowSidebar(IsPanelActive(), true);
 }
 
 void SidebarContainerView::UpdateSidebar() {
@@ -184,10 +184,7 @@ void SidebarContainerView::OnThemeChanged() {
   UpdateBackground();
 }
 
-bool SidebarContainerView::ShouldShowSidebar() const {
-  // Always show if panel should be visible. It is more reliable to check
-  // whether the active index is a panel item rather than checking if
-  // side_panel_ is visible.
+bool SidebarContainerView::IsPanelActive() const {
   auto active_index = sidebar_model_->active_index();
   bool panel_is_active = false;
   if (active_index != -1) {
@@ -197,8 +194,14 @@ bool SidebarContainerView::ShouldShowSidebar() const {
       panel_is_active = true;
     }
   }
+  return panel_is_active;
+}
 
-  return panel_is_active ||
+bool SidebarContainerView::ShouldShowSidebar() const {
+  // Always show if panel should be visible. It is more reliable to check
+  // whether the active index is a panel item rather than checking if
+  // side_panel_ is visible.
+  return IsPanelActive() ||
          sidebar_control_view_->IsItemReorderingInProgress() ||
          sidebar_control_view_->IsBubbleWidgetVisible();
 }
diff --git a/browser/ui/views/sidebar/sidebar_container_view.h b/browser/ui/views/sidebar/sidebar_container_view.h
index f0a2180915..7e36fd4f9b 100644
--- a/browser/ui/views/sidebar/sidebar_container_view.h
+++ b/browser/ui/views/sidebar/sidebar_container_view.h
@@ -98,6 +98,7 @@ class SidebarContainerView
   void ShowOptionsEventDetectWidget(bool show);
   void ShowSidebar(bool show_sidebar, bool show_event_detect_widget);
   SidebarShowOptionsEventDetectWidget* GetEventDetectWidget();
+  bool IsPanelActive() const;
   bool ShouldShowSidebar() const;
   void UpdateToolbarButtonVisibility();
 
```

petemill replied:
Hmm, I don't love the `SetSidebarShowOption` function having to know under which circumstance the bar should be hidden immediately (i.e. `IsPanelActive`) - seems better if that's handled more centrally. I'm looking to see if we can get an event when the context menu is closed, and call `ShowSidebar(ShouldShowSidebar())`

simonhong commented on browser/ui/sidebar/sidebar_unittest.cc:
```diff
@@ -64,36 +64,36 @@ class SidebarModelTest : public testing::Test, public SidebarModel::Observer {
 
 TEST_F(SidebarModelTest, ItemsChangedTest) {
-  EXPECT_EQ(0UL, model()->data_.size());
+  EXPECT_EQ(4UL, service()->items().size());
```
+1 for browser test cases also. Current browser test code has similar code also.

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/chromium-side-panel/45e5419e2c160c5a7cef1c0e5735a4329f2d670e/storybook/index.html)

petemill commented on browser/ui/sidebar/sidebar_model.cc:
```diff
@@ -108,32 +105,30 @@ void SidebarModel::OnItemAdded(const SidebarItem& item, int index) {
-
-  NOTREACHED();
+  int new_active_index = -1;
+  if (active_index_ == from) {
+    new_active_index = to;
+  }
+  if (active_index_ == to) {
+    new_active_index = (to < from) ? active_index_ + 1 : active_index_ - 1;
+  }
+  if (from > active_index_ && to < active_index_) {
+    new_active_index = active_index_ + 1;
+  }
+  UpdateActiveIndexAndNotify(new_active_index);
```
thanks - for some reason this was still working fine though 🤷

petemill commented on browser/ui/sidebar/sidebar_model.h:
```diff
@@ -81,15 +78,9 @@ class SidebarModel : public SidebarService::Observer,
-  bool IsSidebarWebContents(const content::WebContents* web_contents) const;
-
   // Don't cache item list. list can be changed during the runtime.
   const std::vector<SidebarItem> GetAllSidebarItems() const;
```
I don't believe this changed in this PR. It was part of @simonhong 's initial sidebar implementation. Perhaps we can fix it if we want to in a follow-up.

petemill replied:
Actually, I decided to change this as suggested, due to performance

petemill commented on browser/ui/views/sidebar/sidebar_container_view.cc:
```diff
@@ -205,20 +172,33 @@ gfx::Size SidebarContainerView::CalculatePreferredSize() const {
+  // Always show if panel should be visible. It is more reliable to check
+  // whether the active index is a panel item rather than checking if
+  // side_panel_ is visible.
+  auto active_index = sidebar_model_->active_index();
+  bool panel_is_active = false;
+  if (active_index != -1) {
```
Interesting, will have to experiment with how readable it is! I wonder for complex cases if it's too much contained in 1 line - we will see!

petemill commented on browser/ui/views/side_panel/brave_side_panel.h:
```diff
@@ -0,0 +1,39 @@
+  void ChildVisibilityChanged(View* child) override;
+  void OnThemeChanged() override;
+
+  // views::ViewObserver:
+  void OnChildViewAdded(View* observed_view, View* child) override;
+  void OnChildViewRemoved(View* observed_view, View* child) override;
```
When I tried this, it seemed that checking if any child was visible during the event would get an "incorrect" result. The timing may be different.

```c++
void BraveSidePanel::ViewHierarchyChanged(
      const views::ViewHierarchyChangedDetails& details) {
  // When a child is added or removed, visibility condition must be checked.
  if (details.child != this) {
    UpdateVisibility();
  }
}
```

petemill commented on browser/ui/views/sidebar/sidebar_container_view.cc:
```diff
@@ -337,5 +343,58 @@ void SidebarContainerView::StopBrowserWindowEventMonitoring() {
 }
 
+void SidebarContainerView::OnEntryShown(SidePanelEntry* entry) {
+  is_side_panel_event_ = true;
```
Thank you! I was trying to remember what this was called!

petemill commented on components/sidebar/sidebar_item.cc:
```diff
@@ -8,20 +8,29 @@
                                 BuiltInItemType built_in_item_type,
                                 bool open_in_panel) {
   SidebarItem item;
-  item.url = url;
```
Good consideration for follow-up!

petemill commented on common/extensions/api/_api_features.json:
```diff
@@ -48,8 +47,8 @@
       "chrome://bookmarks/*",
-      "chrome://read-later.top-chrome/*",
-      "chrome://sidebar-bookmarks.top-chrome/*"
+      "chrome://bookmarks-side-panel.top-chrome/*",
```
I believe this is still needed for this page to use the `chrome.bookmarks.*` APIs, unless you know of a reason we can delete it? `chrome://bookmarks-side-panel.top-chrome` is the URL for the chromium side panel which shows bookmarks.

simonhong replied:
Oh, I think we can delete whole `bookmarkManagerPrivate` props above as it doesn't have our own specific matches anymore. Also from `feature_compiler_helper.py`

petemill commented:
@sangwoo108 @mkarolin @simonhong thanks for the reviews - feedback is addressed again. Changes can be seen here: https://github.com/brave/brave-core/compare/45e5419e2c160c5a7cef1c0e5735a4329f2d670e..b1d5662ca9a6b90e74d794f16577462fe0bb18e2

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/chromium-side-panel/b1d5662ca9a6b90e74d794f16577462fe0bb18e2/storybook/index.html)

simonhong commented on components/sidebar/sidebar_service.cc:
```diff
@@ -63,34 +76,22 @@ SidebarItem GetBuiltInItemForType(SidebarItem::BuiltInItemType type) {
   return SidebarItem();
 }
 
-SidebarItem::BuiltInItemType GetBuiltInItemTypeForURL(const std::string& url) {
```
Thanks for deleting obsolete functions!

simonhong reviewed (approved):
++ it seems only network audit test should be fixed :)

petemill commented:
Updated to fix network audit test, review changes can now be seen at https://github.com/brave/brave-core/compare/45e5419e2c160c5a7cef1c0e5735a4329f2d670e..4ffd577477

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/chromium-side-panel/4ffd57747776b5a96e2173d8c1d6fc4bd205c5f7/storybook/index.html)

sangwoo108 commented on browser/ui/sidebar/sidebar_model.cc:
```diff
@@ -184,64 +176,30 @@ void SidebarModel::SetActiveIndex(int index, bool load) {
   const auto iter =
-      std::find_if(items.begin(), items.end(),
-                   [item](const auto& i) { return item.url == i.url; });
+      base::ranges::find_if(items.begin(), items.end(), [&item](const auto& i) {
```
When you use `ranges`, I think you can omit begin(), end() of container.
```suggestion
      base::ranges::find_if(items, [&item](const auto& i) {
```

sangwoo108 commented on components/sidebar/sidebar_service.cc:
```diff
@@ -134,6 +144,98 @@ void SidebarService::MigrateSidebarShowOptions() {
+    }
+    // Remember not to hide this item
+    auto iter = base::ranges::find_if(
+        built_in_items_to_hide.begin(), built_in_items_to_hide.end(),
```
ditto:
```suggestion
        built_in_items_to_hide,
```

sangwoo108 commented on components/sidebar/sidebar_service.cc:
```diff
@@ -203,11 +331,12 @@ std::vector<SidebarItem> SidebarService::GetHiddenDefaultSidebarItems() const {
-                                      added_item.built_in_item_type;
-                             });
+    auto iter =
+        base::ranges::find_if(default_items.begin(), default_items.end(),
```
ditto:

```suggestion
        base::ranges::find_if(default_items,
```

sangwoo108 commented on components/sidebar/sidebar_service.cc:
```diff
@@ -225,75 +354,134 @@ SidebarService::ShowSidebarOption SidebarService::GetSidebarShowOption() const {
+  absl::optional<SidebarItem> default_item;
+  for (const auto& type : *preferred_item_types) {
+    auto found_item_iter = base::ranges::find_if(
+        items_.begin(), items_.end(),
```
ditto:
```suggestion
        items_,
```

sangwoo108 commented on browser/ui/views/sidebar/sidebar_container_view.h:
```diff
@@ -71,19 +78,35 @@ class SidebarContainerView
   void OnActiveIndexChanged(int old_index, int new_index) override;
+  void OnItemRemoved(int index) override;
+
+  // SidePanelEntryObserver:
```
How about having base::ScopedObservation<SidePanelEntry, SidePanelEntryObserver> as member variable for this?

I think there's no RemoveObserver() call for SidebarPanelEntry and that could be guaranteed by this RAII object.

sangwoo108 commented on browser/ui/views/sidebar/sidebar_container_view.h:
```diff
@@ -71,19 +78,35 @@ class SidebarContainerView
+
+  // SidePanelRegistryObserver:
+  void OnEntryRegistered(SidePanelEntry* entry) override;
+  void OnEntryWillDeregister(SidePanelEntry* entry) override;
```
And how about having base::ScopedMultiSourceObservation? For similar reason.

sangwoo108 commented on browser/ui/views/sidebar/sidebar_control_view.h:
```diff
@@ -85,6 +93,7 @@ class SidebarControlView : public views::View,
   void UpdateSettingsButtonState();
   void UpdateBackgroundAndBorder();
 
+  raw_ptr<Delegate> delegate_;
```
nit: for consistency

```suggestion
  raw_ptr<Delegate> delegate_ = nullptr;
```

sangwoo108 commented on components/sidebar/sidebar_service.cc:
```diff
@@ -63,34 +76,22 @@ SidebarItem GetBuiltInItemForType(SidebarItem::BuiltInItemType type) {
-
-SidebarItem GetBuiltInItemForURL(const std::string& url) {
-  return GetBuiltInItemForType(GetBuiltInItemTypeForURL(url));
+std::vector<SidebarItem::BuiltInItemType> GetDefaultBuiltInItemTypes() {
```
How about returning const& in order to avoid copying the entire vector?
```suggestion
const std::vector<SidebarItem::BuiltInItemType>&  GetDefaultBuiltInItemTypes() {
```

petemill replied:
thanks for this nuance!

sangwoo108 commented on components/sidebar/sidebar_service.cc:
```diff
@@ -225,75 +354,134 @@ SidebarService::ShowSidebarOption SidebarService::GetSidebarShowOption() const {
+  // Add the items the user has never seen (or never persisted).
+  // Get the initial order of items so that we can attempt to
+  // insert at the intended order.
+  auto default_item_types = GetDefaultBuiltInItemTypes();
```
If GetDefaultBuiltInItemTypes returns const&, const auto& will be better.
```suggestion
  const auto& default_item_types = GetDefaultBuiltInItemTypes();
```

sangwoo108 reviewed (approved):
LGTM with nits 👍

petemill commented:
@sangwoo108 I addressed your welcome nits via https://github.com/brave/brave-core/compare/4ffd577477..650d1c300b - thanks!

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/chromium-side-panel/650d1c300b965e503ac9863f256df02cb1683e47/storybook/index.html)

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/chromium-side-panel/fd2a6b3bb1bd54d1e70593a4bd1194ca16d78654/storybook/index.html)
